app/javascript/components/server-components/ProductEditPage.tsx

```diff
@@ -109,8 +111,25 @@ const createContextValue = (props: Props) => ({
   cancellationDiscountsEnabled: props.cancellation_discounts_enabled,
 });
 
+const pagesHaveSameContent = (pages1: Page[], pages2: Page[]): boolean =>
+  pages1.length === pages2.length && pages1.every((page, i) => isEqual(page.description, pages2[i]?.description));
+
+const checkIfContentDirty = (product: Product, lastSavedProduct: Product) => {
```

Comment:
Just calling it out again here that we should identify the specific variants that were changed. e.g.

- Product has Variant 1 and & Variant 2
- Alice bought 1 and Bob bought 2
- Variant 2 was updated -> we should only send email to Bob

Could you add a E2E test for:

- Banner is shown for text & file changes
- Email is filtered to the specific variants